torch/distributed/optim/apply_optimizer_in_backward.py

```diff
@@ -0,0 +1,76 @@
+from typing import Any, Dict, Iterable, Type
+
+import torch
+
+__all__ = []  # type: ignore[var-annotated]
+
+
+def _apply_optimizer_in_backward(
+    optimizer_class: Type[torch.optim.Optimizer],
+    params: Iterable[torch.nn.Parameter],
+    optimizer_kwargs: Dict[str, Any],
+) -> None:
+    """
+    Upon backwards(), parameters will fire the corresponding optimizer.
+
+    Note - gradients for these parameters will be set to None after backwards().
+    This means that any other (non applied) optimizer over this parameter will be
+    a no-op.
+
+    Args:
+        optimizer_class: Type[torch.optim.Optimizer]: Optimizer to apply to parameter
+        params: Iterator[nn.Parameter]: parameters to apply optimizer state to
+        optimizer_kwargs: Dict[str, Any]: kwargs to pass to optimizer constructor
+
+    Example::
+        params_generator = model.parameters()
+        param_1 = next(params_generator)
+        remainder_params = list(params_generator)
+
+        apply_optimizer_in_backward(torch.optim.SGD, [param_1], {"lr": .02})
+        apply_optimizer_in_backward(torch.optim.Adam, remainder_params, {"lr": .04})
+
+        model(...).sum().backward() # after backward, parameters will already
+        # have their registered optimizer applied.
+
+    """
+
+    def _apply_optimizer_in_backward_to_param(param: torch.nn.Parameter) -> None:
```

Comment:
Do we want to call it `_apply_optimizer_in_backward_to_param()` or simply just `_apply_optimizer_in_backward()`?

I might have missed some naming discussion. If we go with the former, then we should update the docstring example.

Reply:
The overall function is called _apply_optimizer_in_backward, this is an inner helper method that applies optimizer to a single param.

Reply:
Ah, makes sense!